### ISSUE TITLE
CA1812 warning when using internal type as type parameter

### ISSUE THREAD

**Issue by lonix1** [dotnet/roslyn-analyzers#6086]
### Analyzer

**Diagnostic ID**: [CA1812](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/CA1812): `Avoid uninstantiated internal classes`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 6 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 6.0.302](https://dotnet.microsoft.com/download/dotnet/6.0)

### Describe the bug

This was already discussed in #3210 and closed prematurely. There have been other reports since it was closed, but has not been reopened.

### Steps To Reproduce

- I have a client library
- inside there is an `internal` type `Foo`
- and inside that library, it adds that type to the ASP.NET Core IoC container: `services.AddHostedService<Foo>()`
- that older issue also documents other ways this could happen

...but the analyser still warns with `CA1812`

### Expected behavior
Should not show CA1812

### Actual behavior
Shows CA1812

## Additional context
-

**Comment by geeknoid** [comment]
As I mentioned in https://github.com/dotnet/roslyn-analyzers/issues/3210, this diagnostic is basically incompatible with private types being used with DI. We get hundreds of spurious warnings in our source base, which is a shame since this is a useful rule to find dead code.

**Comment by pinkfloydx33** [comment]
Just removed IVT and got spammed with warnings for types used via DI and registered using generic extension methods. Please fix. I really don't want to suppress everything as it's a useful diagnostic otherwise and I'd like to remove IVT from shipped binaries for various reasons.

**Comment by znakeeye** [comment]
Yes, **this needs to be fixed**! I started off making my classes `public` (bad idea, I know) just to avoid this warning.

Consider the sample below. Here, `MyInternalFoo` will produce a `CA1812`. Well, too bad I have a gazillion of these registrations. Same problem with pretty much every DI container out there.

```c#
public static void ConfigureServices(IServiceCollection services)
{
    services.AddSingleton<IFoo, MyInternalFoo>();
}
```

**Comment by Youssef1313** [comment]
Duplicate of https://github.com/dotnet/roslyn-analyzers/issues/4025
